packages/next/src/build/webpack/plugins/slow-module-detection-plugin.ts

```diff
@@ -0,0 +1,255 @@
+      throw new Error(
+        `Invariant: Module is recorded after the report is generated. This is a Next.js internal bug.`
+      )
+    }
+
+    if (duration < this.slowModuleThresholdMs) {
+      return // Skip fast modules
+    }
+
+    getModuleBuildInfo(module).slowModuleDetectionTiming = duration
+    this.pendingModules.push(module)
+  }
+
+  /**
+   * For each slow module, traverses up the dependency chain to find all ancestor modules.
+   * Builds a directed graph where:
+   * 1. Each slow module and its ancestors become nodes
+   * 2. Edges represent "imported by" relationships
+   * 3. Root nodes are entry points with no parents
+   *
+   * The resulting graph allows us to visualize the import chains that led to slow builds.
+   */
+  private prepareReport(compilation: Compilation) {
+    for (const module of this.pendingModules) {
+      // Track visited modules to detect circular dependencies
+      const chain: Module[] = []
+      const visitedModules = new Set<Module>()
+      let currentModule = module
+
+      chain.push(currentModule)
+      visitedModules.add(currentModule)
+
+      // Walk up the module graph until we hit a root module (no issuer)
+      while (true) {
+        const issuerModule = compilation.moduleGraph.getIssuer(currentModule)
+        if (!issuerModule) break
+        if (visitedModules.has(issuerModule)) {
+          throw new Error(
+            `Invariant: Circular dependency detected in module graph. This is a Next.js internal bug.`
+          )
+        }
+        chain.push(issuerModule)
+        visitedModules.add(issuerModule)
+        currentModule = issuerModule
+      }
+
+      // Add all modules in the chain to our graph
+      const modulesInChain: Module[] = []
+      for (const moduleInChain of chain) {
+        const moduleId = getModuleIdentifier(moduleInChain)
+        if (!this.modules.has(moduleId)) {
+          this.modules.set(moduleId, moduleInChain)
+        }
+        modulesInChain.push(moduleInChain)
+      }
+
+      // Create parent-child relationships between adjacent modules in the chain
+      for (let i = 0; i < modulesInChain.length - 1; i++) {
+        const child = modulesInChain[i]
+        const parent = modulesInChain[i + 1]
+
+        this.moduleParents.set(child, parent)
+
+        let parentChildren = this.moduleChildren.get(parent)
+        if (!parentChildren) {
+          parentChildren = new Map()
+          this.moduleChildren.set(parent, parentChildren)
+        }
+        parentChildren.set(getModuleIdentifier(child), child)
+      }
+    }
+
+    this.isFinalized = true
+  }
+
+  generateReport(compilation: Compilation) {
+    if (!this.isFinalized) {
+      this.prepareReport(compilation)
+    }
+
+    // Find root modules (those with no parents)
+    const rootModules = [...this.modules.values()].filter(
+      (node) => !this.moduleParents.has(node)
+    )
+
+    // Truncates long paths by keeping start and end portions
+    const truncatePath = (
```

Comment:
can we move this to a function outside of the class and either have some tests or comment of what it's doing?
e.g. 
```
input: xxx 
output: xxx
```

We might forget the details soon after few months, this would be helpful for this magical formatting